src/storage/mvcc/reader/forward_scanner.rs

```diff
@@ -0,0 +1,443 @@
+                    }
+                    (Some(wk), Some(lk)) => {
+                        let write_user_key = Key::truncate_ts_for(wk)?;
+                        match write_user_key.cmp(lk) {
+                            Ordering::Less => {
+                                // Write cursor user key < lock cursor, it means the lock of the
+                                // current key that write cursor is pointing to does not exist.
+                                (write_user_key, true, false)
+                            }
+                            Ordering::Greater => {
+                                // Write cursor user key > lock cursor, it means we got a lock of a
+                                // key that does not have a write. In SI, we need to check if the
+                                // lock will cause conflict.
+                                (lk, false, true)
+                            }
+                            Ordering::Equal => {
+                                // Write cursor user key == lock cursor, it means the lock of the
+                                // current key that write cursor is pointing to *exists*.
+                                (lk, true, true)
+                            }
+                        }
+                    }
+                };
+
+                // Use `from_encoded_slice` to reserve space for ts, so later we can append ts to
+                // the key or its clones without reallocation.
+                (Key::from_encoded_slice(res.0), res.1, res.2)
+            };
+
+            // Attempt to read specified version of the key. Note that we may get `None`
+            // indicating that no desired version is found, or a DELETE version is found, or
+            // when `has_write == false`. We may also get `Err` due to key lock. We need to
+            // ensure that the scanner keeps working for future calls even when meeting key
+            // lock errors (why??), so we don't apply `?` operator to the result here thus
+            // iterators can be moved forward.
+            let result = self.get(&current_user_key, has_write, has_lock);
+
+            if has_write {
+                self.move_write_cursor_to_next_user_key(&current_user_key)?;
+            }
+            if has_lock {
+                self.lock_cursor.next(&mut self.statistics.lock);
+            }
+
+            // If we got something, it can be just used as the return value. Otherwise, we need
+            // to continue stepping the cursor.
+            if let Some(v) = result? {
+                return Ok(Some((current_user_key, v)));
+            }
+        }
+    }
+
+    /// Attempt to get the value of a key specified by `user_key` and `self.ts`. This function
+    /// requires that the write cursor is currently pointing to the latest version of `user_key`.
+    #[inline]
+    fn get(&mut self, user_key: &Key, has_write: bool, has_lock: bool) -> Result<Option<Value>> {
```

Comment:
You can return if we met next user key, so we can skip move_to_next_user_key for write cursor, this will reduce the calling of `valid()`, `key()` and `is_user_key_eq()`. So the get function is like this:
```
fn get(&mut self, user_key: &Key, met_next_user_key: &mut bool)
```